### PR TITLE
Separate hashchain logic into run_hashchain_for_round() (#33)

### PR DESCRIPTION
**Pull request by alexander-zw**
See #33

Other changes include adding documentation, fixing bugs, and refactoring functions and loops

### CODE REVIEW

File: libiop/bcs/bcs_prover.tcc [hunk 1]
```diff
@@ -35,23 +35,18 @@ void bcs_prover<FieldT, MT_hash_type>::signal_prover_round_done()
      */
     for (auto &kv : mapping)
     {
-        std::vector<std::shared_ptr<std::vector<FieldT>>> all_evaluated_contents;
-        for (auto &v : kv.second)
+        std::vector<std::shared_ptr<std::vector<FieldT>>> all_oracle_evaluated_contents;
+        for (auto &v : kv.second) // kv.second is the oracle handlex.
```
Comment (ValarDragon) [inline libiop/bcs/bcs_prover.tcc:39]:
```suggestion
        for (auto &v : kv.second) // kv.second is a vector of all oracle handles over this domain from this round.
```

File: libiop/bcs/bcs_verifier.tcc [hunk 2]
```diff
@@ -57,19 +58,35 @@ void bcs_verifier<FieldT, hash_digest_type>::seal_interaction_registrations()
...
-        /* Absorb MT roots */
-        for (auto &kv : mapping)
+        /* Absorb MT roots into hashchain. */
+        const std::vector<hash_digest_type> MT_roots_for_round(
+            this->transcript_.MT_roots_.cbegin() + processed_MTs,
+            this->transcript_.MT_roots_.cbegin() + processed_MTs + num_oracles);
+        this->run_hashchain_for_round(round, MT_roots_for_round, this->transcript_.prover_messages_);
+
+        /* Validate all MT queries relative to the transcript. */
+        for (std::size_t i = 0; i < num_oracles; i++)
```
Comment (ValarDragon) [inline libiop/bcs/bcs_verifier.tcc:68]: This loop should be ran only once for every processed MT

Comment (ValarDragon) [reply libiop/bcs/bcs_verifier.tcc:68]: (So once per domain, not once per oracle) And the processed MT counter should be incrementing based off of this

File: libiop/bcs/bcs_prover.tcc [hunk 3]
```diff
@@ -78,21 +73,27 @@ void bcs_prover<FieldT, MT_hash_type>::signal_index_submissions_done()
         throw std::invalid_argument("Didn't provide prover index to BCS prover");
     }
     iop_protocol<FieldT>::signal_prover_round_done();
-    std::size_t ended_round = this->num_prover_rounds_done_-1;
-    const domain_to_oracles_map mapping = this->oracles_in_round(ended_round);
 
-    /* First, go through all the oracle messages in this round and
-       fill in the merkle tree from the index. */
-    for (auto &kv : mapping)
```
Comment (ValarDragon) [inline libiop/bcs/bcs_prover.tcc:86]: We still have to loop through all the different domains in a round

File: libiop/bcs/bcs_verifier.tcc (hunk 2 above)
Comment (alexander-zw) [reply libiop/bcs/bcs_verifier.tcc:68]: It's actually doing that, I just named the variables wrong

File: libiop/bcs/bcs_prover.tcc (hunk 3 above)
Comment (alexander-zw) [reply libiop/bcs/bcs_prover.tcc:86]: It is doing so in `bcs_prover::run_hashchain_for_round`

**Pull request merged**